### PR TITLE
feat(runtask, CAST): add update upgrade result template function

### PR DESCRIPTION
**Pull request by sagarkrsd**
**What this PR does / why we need it**:
This template function will be used to update an upgrade result instance 
with the details of a current runtask execution i.e. the status of the 
runtask execution whether it is successful or has failed,  the message having
details of success/ failure, etc.

An example runtask using this template function to update an upgrade result instance is -

```yaml
# This runtask will patch the upgrade result CR
# with the basic details such as name, namespace and kind
apiVersion: openebs.io/v1alpha1
kind: RunTask
metadata:
  name: upgrade-cstor-volume-080-081-patch-upgrade-result
  namespace: default
spec:
  meta: |
    id: patchResult
    apiVersion: openebs.io/v1alpha1
    kind: UpgradeResult
    action: patch
    objectName: {{ .UpgradeItem.upgradeResultName }}
    runNamespace: default
  post: |
    {{ $retries := 0 -}}
    {{- $message := printf "upgradeResult {%s} has been patched with basic details such as name and namespace of the resource to be upgraded." .UpgradeItem.upgradeResultName -}}
    {{- $URName := upgradeResultWithTaskOwnerName .UpgradeItem.upgradeResultName -}}
    {{- $URNamespace := upgradeResultWithTaskOwnerNamespace "default" -}}
    {{- $taskName := upgradeResultWithTaskName "upgrade-cstor-volume-080-081-patch-upgrade-result" -}}
    {{- $taskStatus := upgradeResultWithTaskStatus "successful" -}}
    {{- $taskMessage := upgradeResultWithTaskMessage $message -}}
    {{- $taskStartTime := upgradeResultWithTaskStartTime now -}}
    {{- $taskEndTime := upgradeResultWithTaskEndTime now -}}
    {{- $taskRetries := upgradeResultWithTaskRetries 7 -}}
    {{- upgradeResultUpdateTasks $taskStartTime $taskRetries $URName $URNamespace $taskName $taskStatus $taskMessage $taskEndTime -}}
  task: |-
    type: merge
    pspec: |-
      status:
        resource:
          name: {{ .UpgradeItem.name }}
          namespace: {{ .UpgradeItem.namespace }}
          kind: {{ .UpgradeItem.kind }}

```
Signed-off-by: sagarkrsd <sagar.kumar@openebs.io>

### CODE REVIEW

File: pkg/task/task.go [hunk 1]
```diff
@@ -1423,8 +1423,12 @@ func (m *executor) putCStorVolumeReplica() (err error) {
 
 // putUpgradeResult will put an upgrade result as defined in the task
 func (m *executor) putUpgradeResult() (err error) {
+	raw, err := template.AsTemplatedBytes("UpgradeResult", m.Runtask.Spec.Task, m.Values)
+	if err != nil {
+		return
+	}
 	uresult, err := upgraderesult.
-		BuilderForRuntask("UpgradeResult", m.Runtask.Spec.Task, m.Values).
+		BuilderForTemplateObject(raw).
```
Comment (AmitKumarDas) [inline pkg/task/task.go]: Can this be `BuilderForYAMLObject`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 2]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	for _, o := range opts {
+		o := o
+		o(u)
+	}
+	return u
+}
+
+// Update is a template function exposed for
+// updating an upgrade result instance
+func Update(opts ...UpgradeResultUpdateOption) error {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: can we rename this to `UpdateTasks`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 3]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	u := NewUpdateUpgradeResult(opts...)
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: rename `u` to `new`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 4]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	if u.name == "" {
+		return errors.New("missing upgrade result name")
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]:
```go
errors.New("failed to update upgrade result tasks: missing upgrade result name")
```

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 5]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	}
+	// First get the desired upgrade result instance
+	k := KubeClient()
+	k.namespace = u.namespace
+	ur, err := k.Get(u.name, metav1.GetOptions{})
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: change `ur` to `existing`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 6]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	if err != nil {
+		return err
+	}
+	// Iterate over the upgrade result tasks to check if the
+	// desired task to be updated exists or not,
+	// if exists then update the task instance with the given values.
+	for i, task := range ur.Tasks {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: change `task` to `existingTask`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 7]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+		if task.Name == u.task.Name {
+			task = *u.task
+		}
+		ur.Tasks[i] = task
+	}
+	// Update the upgrade result instance with
+	// the provided values
+	_, err = k.Update(ur)
+	if err != nil {
+		return err
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]:
```go
  return errors.Wrapf(
    err, 
    "failed to update upgrade result tasks: name {%s} namespace {%s}", 
    ur.Name, 
    ur.Namespace,
  )
```

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 8]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	u := NewUpdateUpgradeResult(opts...)
+	if u.name == "" {
+		return errors.New("missing upgrade result name")
+	}
+	// First get the desired upgrade result instance
+	k := KubeClient()
+	k.namespace = u.namespace
+	ur, err := k.Get(u.name, metav1.GetOptions{})
+	if err != nil {
+		return err
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]:
```go
  return errors.Wrapf(
    err, 
    "failed to update upgrade result tasks: name {%s} namespace {%s}", 
    ur.Name, 
    ur.Namespace,
  )
```

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 9]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	}
+	return nil
+}
+
+// TemplateFunctions exposes a few functions as
+// go template functions to be used for upgrade result
+func TemplateFunctions() template.FuncMap {
+	return template.FuncMap{
+		"updateUpgradeResult": Update,
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: rename to "upgradeResultUpdateTasks"

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 10]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+		"withURName":          WithURName,
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: rename to "upgradeResultWithName"

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 11]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+		"withURNamespace":     WithURNamespace,
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: change other names similarly

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 12]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	cs, err := k.getClientOrCached()
+	if err != nil {
+		return nil, err
+	}
+	return k.update(cs, updateObj, k.namespace)
+}
+
+// UpdateUpgradeResult enables update
+// operation on upgrade result instance
+type UpdateUpgradeResult struct {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: Let us reuse the existing custom `UpgradeResult`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 13]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	name      string
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: Add these to custom `UpgradeResult` struct

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 14]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	namespace string
+	task      *apis.UpgradeResultTask
+}
+
+// UpgradeResultUpdateOption defines the abstraction
+// to build an update instance for upgrade result
+type UpgradeResultUpdateOption func(*UpdateUpgradeResult)
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: Let us reuse the existing `UpgradeResultBuildOption`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 15]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+
+// WithURName sets the name of the upgrade
+// result
+func WithURName(name string) UpgradeResultUpdateOption {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: `WithName`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 16]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+	return func(u *UpdateUpgradeResult) {
+		u.name = name
+	}
+}
+
+// WithURNamespace sets namespace where upgrade
+// result is present
+func WithURNamespace(namespace string) UpgradeResultUpdateOption {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: `WithNamespace`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 17]
```diff
@@ -210,3 +229,153 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+// a runtask has retried executing a particular task
+func WithTaskRetries(retries int) UpgradeResultUpdateOption {
+	return func(u *UpdateUpgradeResult) {
+		u.task.Retries = retries
+	}
+}
+
+// NewUpdateUpgradeResult returns a new instance of updateUpgradeResult
+// meant for updating an upgrade result instance
+func NewUpdateUpgradeResult(opts ...UpgradeResultUpdateOption) *UpdateUpgradeResult {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: Let us reuse the existing constructor

File: pkg/upgrade/result/v1alpha1/upgraderesult.go [hunk 18]
```diff
@@ -95,21 +94,16 @@ func (b *Builder) WithResultConfig(resource apis.ResourceDetails,
 	return b
 }
 
-// BuilderForRuntask returns a new instance
-// of Builder for runtasks
-func BuilderForRuntask(context, templateYaml string, templateValues map[string]interface{}) *Builder {
+// BuilderForTemplateObject returns a new instance
+// of Builder for a given template object
+func BuilderForTemplateObject(object []byte) *Builder {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/upgraderesult.go]: `BuilderForYAMLObject`

File: pkg/task/task.go (hunk 1 above)
Comment (sagarkrsd) [reply pkg/task/task.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 2 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 3 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 4 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 5 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 8 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 7 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 6 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 9 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 10 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 11 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/upgraderesult.go (hunk 18 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/upgraderesult.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 16 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 15 above)
Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

File: pkg/upgrade/result/v1alpha1/kubernetes.go (hunk 12 above)
Comment (AmitKumarDas) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Can we name this as `UpgradeResultForTask`

File: pkg/upgrade/result/v1alpha1/kubernetes.go [hunk 19]
```diff
@@ -210,3 +228,163 @@ func (k *kubeclient) Patch(name string, pt types.PatchType,
...
+// a runtask has retried executing a particular task
+func WithTaskRetries(retries int) UpgradeResultForTaskOption {
+	return func(u *UpgradeResultForTask) {
+		u.task.Retries = retries
+	}
+}
+
+// NewUpdateUpgradeResult returns a new instance of updateUpgradeResult
+// meant for updating an upgrade result instance
+func NewUpdateUpgradeResult(opts ...UpgradeResultForTaskOption) *UpgradeResultForTask {
```
Comment (AmitKumarDas) [inline pkg/upgrade/result/v1alpha1/kubernetes.go]: `NewUpgradeResultForTask`

Comment (sagarkrsd) [reply pkg/upgrade/result/v1alpha1/kubernetes.go]: Done

**Pull request merged**